src/translator/cache.h

```diff
@@ -0,0 +1,205 @@
+  size_t hashTableIndex_;
+
+  HashCacheKey hashFn_;
+};
+
+/// Alternative cache for non-thread based workflow (specifically WASM). LRU Eviction Policy. Uses a lot of std::list.
+/// Yes, std::list; If someone needs a more efficient version, look into threads and go for ThreadSafeL4Cache.
+/// This class comes as an afterthought, so expect parameters which may not be used to achieve parity with
+/// ThreadSafeL4Cache.
+
+class ThreadUnsafeLRUCache : public TranslationCache {
+ public:
+  /// Construct a ThreadUnSafeLRUCache from configuration passed via key values in an Options object:
+  ///
+  /// @param [in] options: Options object which is used to configure the cache. See command-line parser for
+  /// documentation (`marian::bergamot::createConfigParser()`)
+  ThreadUnsafeLRUCache(const CacheConfig &config);
+
+  /// Fetches a record from cache, storing it in processedRequestSentence if found. Not thread-safe.
+  ///
+  /// @param [in]  words: query marian:Words for which translation results are looked up in cache.
+  /// @param [out] processedRequestSentence: stores cached results for use outside if found.
+  ///
+  /// @returns true if query found in cache false otherwise.
+  bool fetch(const TranslationModel *model, const marian::Words &words,
+             ProcessedRequestSentence &processedRequestSentence);
+
+  /// Inserts a new record into cache.
+  ///
+  /// @param [in] words: marian::Words processed from a sentence
+  /// @param [in] processedRequestSentence: minimum translated information corresponding to words
+  void insert(const TranslationModel *model, const marian::Words &words,
+              const ProcessedRequestSentence &processedRequestSentence);
+
+  CacheStats stats() const;
+
+ private:
+  // A Record type holds Key and Value together in a linked-list as a building block for the LRU cache.
+  // Currently we use uint64_t for Key to reduce collisions. Value (Storage) is simply a binary sequential memory with
+  // some enhancements to read variables and ranges from the memory. The API works with marian::Words (internally
+  // reduced to uint6_t to reduce vector comparisons) and ProcessedRequestSentence (which can be constructed given a
+  // Storage - which is a more compressed representation).
+  struct Record {
+    using Key = size_t;
+    using Value = Storage;
+
+    Key key;
+    Value value;
+
+    size_t size() const { return /*key=*/sizeof(Key) + value.size(); }
+  };
+
+  // A list representing memory where the items are stored and a hashmap (unordered_map) with values pointing to this
+  // storage is used for a naive LRU implementation.
+  using RecordList = std::list<Record>;
+  using RecordPtr = RecordList::iterator;
+
+  RecordList storage_;
+  std::unordered_map<Record::Key, RecordPtr> cache_;
+
+  /// Active sizes (in bytes) stored in storage_
+  size_t storageSize_;
+
+  // Limit of size (in bytes) of storage_
+  size_t storageSizeLimit_;
+
+  HashCacheKey hashFn_;
```

Comment:
Again this is strange, as it holds no state.

Reply:
This is now a static member function at the interface, with CacheKey and hash(CacheKey) being protected members. Anyone outside is thus forbidden from using these, and the static member function should eliminate any problems with "state".